### PR TITLE
Implement add_time_duration_null; Fix month_name; Add more type annotation

### PR DESCRIPTION
**Pull request by koushiro**
Signed-off-by: koushiro <koushiro.cqx@gmail.com>

## What have you changed? (mandatory)

Implement add_time_duration_null;
Fix month_name; 
Add more type annotation.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

unit-test

## Does this PR affect documentation (docs) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @AndreMouche @breeswish

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -32,11 +32,11 @@ impl ScalarFunc {
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, [u8]>>> {
-        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        let t: Cow<'a, Time> = try_opt!(self.children[0].eval_time(ctx, row));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: why we need the type annotation here?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -102,12 +102,12 @@ impl ScalarFunc {
...
         let month = t.get_time().month() as usize;
         if t.is_zero() && ctx.cfg.mode_no_zero_date_mode() {
             return handle_incorrect_datetime_error(ctx, t).map(|_| None);
         } else if month == 0 || t.is_zero() {
-            return Ok(Some(Cow::Owned("".to_string().into_bytes())));
+            return Ok(None);
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Any test cover it?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: Just for supporting auto-completion better in IDE.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: Related test cases have just been added. PTAL again.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_time.rs]: I think you'd better remove it to make the code style consistent.

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: I think the code style is more consistent by adding type annotation in builtin_time.rs, like `year_week_with_mode`, `year_week_without_mode`, `add_datetime_and_duration`, `add_duration_and_duration` and so on.

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: I know the rust compiler can infer type, but macros like `try_opt!` are not friendly to the IDE's or editor's auto-completion.
If you think these type annotations must be removed, I will. But how about functions like `year_week_with_mode`? Remove type annotation of them together?

Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_time.rs]: Thanks for your explanation, it's our fault to make the code style not consistent before. You are right to make it consistent in this file, so we'd like to accept your edition.

Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: @brson how do you think about this, is it a good idea to add type annotation here?

Comment (breeswish) [reply src/coprocessor/dag/expr/builtin_time.rs]: FYI, in future, we don't need it since null will be treated separately. For now, it's fine.

**Pull request merged**